### PR TITLE
Services tests refactoring

### PR DESCRIPTION
**Pull request by ArBond**

### CODE REVIEW

File: Lab08.MVC/Lab08.MVC.WEB.Tests/Services/BaseServiceTests.cs [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+        }
+
+        #region Constructor tests
+        [Test()]
+        public void Constructor_When_UnitOfWorkArgumentIsNull_Should_ThrowsNullArgumentException()
+        {
+            Assert.Throws<ArgumentNullException>(() => new BaseService<long, Mock, Mock>(null, _converter.Object));
+        }
+
+        [Test()]
```
Comment (Glareone) [inline Lab08.MVC/Lab08.MVC.WEB.Tests/Services/BaseServiceTests.cs]: if you don't want to provide any information in parentheses - you could use:
[Test]

File: Lab08.MVC/Lab08.MVC.WEB.Tests/Services/BaseServiceTests.cs [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        public void Constructor_When_ConverterArgumentIsNull_Should_ThrowsNullArgumentException()
+        {
+            Assert.Throws<ArgumentNullException>(() => new BaseService<long, Mock, Mock>(_unitOfWork.Object, null));
+        }
+
+        [Test()]
+        public void Constructor_When_RepositoryIsNull_Should_ThrowsNullArgumentException()
+        {
+            _unitOfWork.Setup(u => u.GetRepository<long, Mock>()).Returns(() => null);
+            Assert.Throws<ArgumentNullException>(() => new BaseService<long, Mock, Mock>(_unitOfWork.Object, _converter.Object));
```
Comment (Glareone) [inline Lab08.MVC/Lab08.MVC.WEB.Tests/Services/BaseServiceTests.cs]: better to check that all is okay with Resolved components (positive tests).
What I mean:

            object target = null;
            Assert.DoesNotThrow(() =>
            {
                target = _unitOfWork.Setup(u => u.GetRepository<long, Mock>());;
            });
            Assert.IsNotNull(target);

but if you would like to test exception type-go ahead with this approach

File: Lab08.MVC/Lab08.MVC.WEB.Tests/Services/BaseServiceTests.cs [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        public void GetAllByPredicate_IsGetAllByPredicateMethodUsedAndReturnsEntitiesCollection()
+        {
+            var expectedList = new List<Mock>();
+            _repository.Setup(r => r.GetAll(It.IsAny<Func<Mock, bool>>())).Returns(expectedList);
+            var service = new BaseService<long, Mock, Mock>(_unitOfWork.Object, _converter.Object);
+            bool func(Mock x) => true;
+
+            var actualList = service.GetAll(func);
+            
+            _repository.Verify(r => r.GetAll(It.IsAny<Func<Mock, bool>>()));
```
Comment (Glareone) [inline Lab08.MVC/Lab08.MVC.WEB.Tests/Services/BaseServiceTests.cs]: if you will check that this method take proper parameters - will be better:
proper predicate.
proper parameters.

 _repository.Verify(
                x => x.GetAll(It.Is<Func<Mock, bool>>(s => s == YOURCLAUSE)), Times.Once());

File: Lab08.MVC/Lab08.MVC.WEB.Tests/Services/JourneyServiceTests.cs [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+using Lab08.MVC.DTO;
+using Moq;
+using NUnit.Framework;
+using System;
+using System.Collections.Generic;
+using System.Data;
+
+namespace Lab08.MVC.BLL.Services.Tests
+{
+    [TestFixture()]
```
Comment (Glareone) [inline Lab08.MVC/Lab08.MVC.WEB.Tests/Services/JourneyServiceTests.cs]: [TestFixture(Category = "Unit Tests")]

File: Lab08.MVC/Lab08.MVC.WEB.Tests/Services/BaseServiceTests.cs [hunk 5]
```diff
@@ -0,0 +1,161 @@
+﻿using System;
+using System.Collections.Generic;
+using Lab08.MVC.BLL.Interfaces;
+using Lab08.MVC.DAL.Interfaces;
+using Moq;
+using NUnit.Framework;
+
+namespace Lab08.MVC.BLL.Services.Tests
+{
+    [TestFixture()]
```
Comment (Glareone) [inline Lab08.MVC/Lab08.MVC.WEB.Tests/Services/BaseServiceTests.cs]: [TestFixture(Category = "Unit Tests")]

File: Lab08.MVC/Lab08.MVC.WEB.Tests/Services/JourneyServiceTests.cs [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+            journeyRouteRepository.Setup(r => r.Get(It.IsAny<long>())).Returns(() => null);
+            _converter.Setup(c => c.DtoToDomain(It.IsAny<JourneyDto>())).Returns(new Journey());
+            _unitOfWork.Setup(u => u.GetRepository<long, User>()).Returns(userRepository.Object);
+            _unitOfWork.Setup(u => u.GetRepository<long, JourneyRoute>()).Returns(journeyRouteRepository.Object);
+
+            Assert.Throws<DataException>(() => _journeyService.Create(new JourneyDto()));
+        }
+
+        [Test()]
+        public void Create_IsCreateAndSaveMethodsUsed()
```
Comment (Glareone) [inline Lab08.MVC/Lab08.MVC.WEB.Tests/Services/JourneyServiceTests.cs]: Create_AreCreateAndSaveMethodsUsed. But better to use Does:
Create_DoesCreateAndSaveMethodWorkWithProperParameters

File: Lab08.MVC/Lab08.MVC.WEB.Tests/Services/JourneyServiceTests.cs [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+        {
+            var userRepository = new Mock<IRepository<long, User>>();
+            var journeyRouteRepository = new Mock<IRepository<long, JourneyRoute>>();
+            journeyRouteRepository.Setup(r => r.Get(It.IsAny<long>())).Returns(new JourneyRoute());
+            _converter.Setup(c => c.DtoToDomain(It.IsAny<JourneyDto>())).Returns(new Journey());
```
Comment (Glareone) [inline Lab08.MVC/Lab08.MVC.WEB.Tests/Services/JourneyServiceTests.cs]: to init stage. You initialize every time on the same manner.

if you want to set parameters to your initialization: create a new method like InitializeConverterWithParameters(parameters) and use it when it needed

File: Lab08.MVC/Lab08.MVC.WEB.Tests/Services/JourneyServiceTests.cs [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+            _unitOfWork.Setup(u => u.GetRepository<long, User>()).Returns(userRepository.Object);
+            _unitOfWork.Setup(u => u.GetRepository<long, JourneyRoute>()).Returns(journeyRouteRepository.Object);
+
+            Assert.Throws<DataException>(() => _journeyService.Create(new JourneyDto()));
+        }
+
+        [Test()]
+        public void Create_IsCreateAndSaveMethodsUsed()
+        {
+            var userRepository = new Mock<IRepository<long, User>>();
```
Comment (Glareone) [inline Lab08.MVC/Lab08.MVC.WEB.Tests/Services/JourneyServiceTests.cs]: instantly initialize this variable on the top of the class:

private Mock<IRepository<long, User>> _userRepository; + init in Init method.

or

private Mock<IRepository<long, User>> _userRepository = new Mock<IRepository<long, User>>();

File: Lab08.MVC/Lab08.MVC.WEB.Tests/Services/JourneyServiceTests.cs [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+            userRepository.Setup(r => r.Get(It.IsAny<long>())).Returns(() => null);
+            _unitOfWork.Setup(u => u.GetRepository<long, User>()).Returns(userRepository.Object);
+
+            Assert.Throws<DataException>(() => _journeyService.Subscribe(0, 0));
+        }
+
+        [Test()]
+        public void Subscribe_IsGetAddToParticipantsUpdateAndSaveMethodsUsed()
```
Comment (Glareone) [inline Lab08.MVC/Lab08.MVC.WEB.Tests/Services/JourneyServiceTests.cs]: best candidate for integration test with real database in memory;
"add new participant"

to use in memory db:

            var options = new DbContextOptionsBuilder<DbContext>()
                    .UseInMemoryDatabase(Guid.NewGuid().ToString())
                    .Options;

            _data = new Data(new DbContext(options));

take a look on examples from internet

File: Lab08.MVC/Lab08.MVC.WEB.Tests/Services/UserServiceTests.cs [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+        }
+
+        private bool UserDtoEquals(UserDto user1, UserDto user2)
+        {
+            if (user1 == null && user2 == null)
+                return true;
+            else if (user1 == null || user2 == null)
+                return false;
+
+            if (user1.UserId == user2.UserId &&
```
Comment (Glareone) [inline Lab08.MVC/Lab08.MVC.WEB.Tests/Services/UserServiceTests.cs]: return user1.UserId == user2.UserId &&
                user1.Login == user2.Login &&
                user1.PasswordHash == user2.PasswordHash &&
                user1.Salt == user2.Salt &&
                user1.FirstName == user2.FirstName &&
                user1.LastName == user2.LastName &&
                user1.Nickname == user2.Nickname &&
                user1.CreationTime == user2.CreationTime &&
                user1.Phone == user2.Phone;

File: Lab08.MVC/Lab08.MVC.WEB.Tests/Services/BaseServiceTests.cs (hunk 1 above)
Comment (ArBond) [reply Lab08.MVC/Lab08.MVC.WEB.Tests/Services/BaseServiceTests.cs]: done

File: Lab08.MVC/Lab08.MVC.WEB.Tests/Services/JourneyServiceTests.cs (hunk 4 above)
Comment (ArBond) [reply Lab08.MVC/Lab08.MVC.WEB.Tests/Services/JourneyServiceTests.cs]: done

File: Lab08.MVC/Lab08.MVC.WEB.Tests/Services/BaseServiceTests.cs (hunk 5 above)
Comment (ArBond) [reply Lab08.MVC/Lab08.MVC.WEB.Tests/Services/BaseServiceTests.cs]: done

File: Lab08.MVC/Lab08.MVC.WEB.Tests/Services/JourneyServiceTests.cs (hunk 6 above)
Comment (ArBond) [reply Lab08.MVC/Lab08.MVC.WEB.Tests/Services/JourneyServiceTests.cs]: done

File: Lab08.MVC/Lab08.MVC.WEB.Tests/Services/JourneyServiceTests.cs (hunk 8 above)
Comment (ArBond) [reply Lab08.MVC/Lab08.MVC.WEB.Tests/Services/JourneyServiceTests.cs]: done

File: Lab08.MVC/Lab08.MVC.WEB.Tests/Services/JourneyServiceTests.cs (hunk 7 above)
Comment (ArBond) [reply Lab08.MVC/Lab08.MVC.WEB.Tests/Services/JourneyServiceTests.cs]: moved them to the class fields and there initialized

**Pull request merged**